### ISSUE TITLE
Throttling Hive splits discovery & assignment

### ISSUE THREAD

**Issue by BenoitHanotte** [prestodb/presto#12495]
Hello,
We have been experiencing issues recently where some extremely large queries (>1M splits) against Hive tables in a cluster over 100 nodes would flood our HDFS namenode with a very high number of simultaneous read operations. This slowed the namenode down significantly as it was struggling to keep up and started impacting other clients.

We are currently investigating a way to throttle the split discovery/assignment for Hive splits, without throttling the rest of the query stages, the aim being to prevent tasks from all opening too many concurrent splits and basically DDoSing the namenode.
One solution we have is to reduce `hive.split-loader-concurrency` from 4 to 1. However, as much as I understand from looking at the code,  this will reduce the parallelism used to list the content of the partitions, and thus indirectly the rate at which splits are discovered, but does not provide guarantees and would likely only be a partial solution as we have seen queries multiply the average load on the namenode by ~10x over the span of a few minutes.
Another solution could be to raise the split size but that would potentially penalize medium sized queries. We could also create a new tier of split size for very large queries (eg: > 200K splits), however that won't prevent the first splits from creating a quick burst of requests to the namenode.

Has anybody already had a look into a potential solution to this? From looking at the code, there does not seem to be a way to limit concurrency only for source stages that read Hive splits independently of the query.
One thing we are experimenting with is to allow throttling the rate and size of the batches returned by the `AsyncQueue` to avoid releasing too many concurrent splits to tasks of the source stage, but we would definitely welcome a better solution if there was one :)

Thanks!

**Comment by nezihyigitbasi** [comment]
AFAIK you can tune two knobs for that, one is `hive.split-loader-concurrency` as you already figured out, the other one is the number of such queries running at the same time, which is a resource group config problem.

I don't see any throttling or rate limiting in the split loading code path, so currently I don't think we have a good way to prevent a potential spike that would happen if a query arrives at the system that enumerates a large number of splits, even with a concurrency of 1.

I think ideally, and if possible, the table that's being scanned should be fixed to have a smaller number of slightly larger splits as having a large number of splits is a bad physical layout for a table for many reasons. Many deployments use a split size that's equivalent to an HDFS block size of 64MB or 128MB.

**Comment by BenoitHanotte** [comment]
Hello @nezihyigitbasi , Thanks for your reply! 

We have tried to reduce the split loader concurrency, but even with a concurrency of 1, the splits can be assigned to tasks so fast that they can easily trigger over 150K read ops/min:
![loader-concurrency](https://user-images.githubusercontent.com/6229788/54909729-27970300-4eeb-11e9-9359-00a7654ff3d5.png)
Ideally we would like to have strong guarantees that bad queries cannot trigger over 60-80K ops/min.

There have indeed been issue in some datasets where small files were not properly consolidated together thus leading to a unreasonably high number of splits. However, as we don't own the pipelines generating these datasets, we would like to make sure that such an issue won't lead to presto making read requests to the namenode at a unually high rate.

We have been experimenting with throttling the `HiveSplitSource`, I pushed a first PR: https://github.com/prestosql/presto/pull/534. 
This idea is to throttle the `borrowBatchAsync` method of the `AsyncQueue` to only release up to `hive.max-splits-per-sec` splits per second to the source stage.
We have tested it and it proved very effective as you can see in the graph below (0 being no throttling):
![splits-throttling](https://user-images.githubusercontent.com/6229788/54909887-7a70ba80-4eeb-11e9-86ad-359728a1827a.png)

If you had a few minutes, I would gladly welcome feedback on whether you think this would be the right place to implement such a throttling and also if you think this is a feature that could reach upstream.

Thanks!

**Comment by nezihyigitbasi** [comment]
> would flood our HDFS namenode with a very high number of simultaneous read operations. 

Can HDFS namenode throttling support (HADOOP-10597) help here?  

> This idea is to throttle the borrowBatchAsync method of the `AsyncQueue` to only release up to hive.max-splits-per-sec splits per second to the source stage.

Did you consider throttling at `BackgroundHiveSplitLoader.HiveSplitLoaderTask.process()`, which is the task generating the splits and putting into the `AsyncQueue`.

**Comment by BenoitHanotte** [comment]
Hello @nezihyigitbasi , thanks for your reply!

I considered throttling the split-enqueuing process, it was actually my first approach, but it didn't seem the best place to do so as it doesn't give strong guarantees on the rate of the split dequeuing and assignment as it (in theory) doesn't prevent to dequeue a large batch at once. The maximum size of the queue then translates as back-pressure for the `SplitLoaderTask`. I could give another try to this approach, do you believe there is a good reason to do it in the `offer()` method (memory maybe)?

Regarding HADOOP-10597, I wasn't aware of it and I can't seem to be able to find references to it in our Hadoop fork (based on the 2.6 branch from a vendor), I will see with our Hadoop team to investigate the impact and feasibility of it.

**Comment by BenoitHanotte** [comment]
Hello @nezihyigitbasi , I have check with our Hadoop team and actually we have had `HADOOP-10597` backported and in place for quite some time but it seems it not sufficient to provide a "fair" distribution of the load between clients. 
From my discussion with them, this mechanism is meant to allow the namenode to not block on the request and instead fail it immediately. However it seems to use the same mechanism as a failover and does not make the client wait and backoff, and thus does not make heavy-hitting clients slow down their query rate.Thus, a large Presto query would still be able to "unfairly" the queue with a large number of requests to be processed, and would impact the latency of other clients.

This shows however that another approach could be to backoff exponentially when such responses are received, but that would also mean that at that point the namenode is already over-loaded. What do you think?

**Comment by wenleix** [comment]
@BenoitHanotte :

Thanks for reporting the issue. I have two thoughts:

- What's the query trigger the issue? Does it contains many UNION ALLs? -- Because the `AsyncQueue` should be able to "pushback" the splits for each individual sources. So it should happen if you have many split sources. However, if it's the case it not clear to me why throttling on each individual sources would help. Does your query use grouped execution? -- this is the only possibility to bypass the "pushback" from `AsyncQueue`. 

- What's the session property `execution_policy` set to? `all-at-once` will schedule all stages at once which might have higher pressure on split discovery. Have you tried `phased` ? 

For history and overview of `HiveSplitSource`, bucketed table and grouped execution, I wrote a document https://docs.google.com/document/d/1WPuA_pyRhJMahdhzgG58qptY1iNKAjJlOgXLFq4MBR4 , and hope it helps to help developers and community to understand the history and help diagnose the root issues.

**Comment by BenoitHanotte** [comment]
Hello @wenleix,
Thanks for your reply and for the document, that explains a lot about the design of the Hive source!

- We have been able to reproduce the issue with a simple `SELECT` over a large number of partitions (2 weeks of hourly-partitioned data) on a table for which the data has been poorly consolidated (many few-MB files on HDFS). We fixed the consolidation of these files, however as we don't own the pipelines generating the data on HDFS, we want to make sure that we are immune to the same problem happening again and leading to Presto making too many concurrent READ request to the namenode. Also, querying months of data could trigger a large amount of splits even with well-consolidated files again.

- the execution policy is `all-at-once` and we should consider changing it to `phased` as it would help for very large queries having multiple sources. However as the issue can be triggered with a simple `SELECT` over a large amount of data we might still want to protect ourselves from it.

- We are using parquet with 256MB block size by default but haven't tuned the split size in presto I believe. I will need to check that it is aligned with the default page size (which should be 128MB) as, as you described, this might trigger useless splits (I hadn't considered that potential issue).

Thanks for the info, I'll investigate all that next week!
